velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "HdfsFileSystem.h"
+#include <hdfs/hdfs.h>
+#include "HdfsReadFile.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/core/Context.h"
+
+namespace facebook::velox::filesystems {
+folly::once_flag hdfsInitiationFlag;
+std::string_view HdfsFileSystem::kScheme("hdfs:/");
+
+class HdfsFileSystem::Impl {
+ public:
+  explicit Impl(const Config* config) {
+    auto endpointInfo = getServiceEndpoint(config);
+    auto builder = hdfsNewBuilder();
+    hdfsBuilderSetNameNode(builder, endpointInfo.host.c_str());
+    hdfsBuilderSetNameNodePort(builder, endpointInfo.port);
+    hdfsClient_ = hdfsBuilderConnect(builder);
```

Comment:
Is there a default timeout for this if the hdfs cluster is down? Is this configurable in the library?
Can you add a test to read when the hdfs cluster is down?

Reply:
You can configure the settings 2 ways. Via the builder construction (passing the seetings in) or via hdfs-client.xml.
The default is to look for that file in the current directory or by `LIBHDFS3_CONF` env variable.

Just so I have it correct, the test you are suggesting is:

1) hdfsBuilderConnect()
2) kill the test cluster
3) hdfsRead()
4) verify correct behaviour

Reply:
1) We must add the test you suggested. 
2) We must also add another test where we create an HDFSFileSystem (i.e invoke hdfsBuilderConnect()) when the hdfs cluster is down.

Reply:
Ok I agree. Good catch.

Reply:
Can you also write a comment in `HdfsFileSystem.cpp` about how to configure libhdfs3 as you described in your comment above. Thanks.

Reply:
The header `HdfsFileSystem.h` is probably a better place for the comment.

Reply:
Will do